### PR TITLE
R markdown templates

### LINKED ISSUE #839: [Feature request] R Markdown: Choose templates
It would be so great if we had an interactive command showing all the available Rmd skeleton templates that users can choose from.

Currently, RStudio IDE has this feature. As you know, all the Rmd templates path has the following structure:

`[pkg_name]/rmarkdown/templates/`

### PR DESCRIPTION
**Pull request by renkun-ken**
# What problem did you solve?

Closes #839 

This PR implements choosing template and create draft rmd from template via `rmarkdown::draft()`.

The templates are obtained via scanning the R package folders so that the information of all templates are written to a temp JSON file. When user picks a template, saveFileDialog will show up and user needs to find a location to save the file.

Some templates have multiple files. If user specify e.g. `draft.Rmd`, multi-file templates will create a new folder `draft` and then put files in it.

## (If you have)Screenshot

https://user-images.githubusercontent.com/4662568/153199351-3838bdfb-a2a5-44c5-be7d-2dbf752e07b6.mp4

## (If you do not have screenshot) How can I check this pull request?

1. "R Markdown: New Draft" will let user choose template and choose a location to save file.
2. Then a rmd file or a folder will be created.
3. The rmd file will be open in editor.

### CODE REVIEW

**Comment by ManuelHentschel** [comment]
I'm not very familiar with rmd templates, but the ones that were found in my installed packages (rmarkdown, reprex) were only single-file documents. In these instances it might feel more natural to just create an unsaved editor with the content of the file, instead of having to save it somewhere.

Would it be feasible to do this for the packages that have `create_dir = false`?

**Comment by renkun-ken** [comment]
Yes. In my initial implementation, I create untitled documents in all cases, but switched to save before creation when I test with `create_dir = true`. Using untitled documents for `create_dir = false` certainly makes good sense, and it is consistent with the behavior of RStudio. I'll implement this soon.

File: src/rmarkdown/draft.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+export async function newDraft(): Promise<void> {
+    const cwd = getCurrentWorkspaceFolder()?.uri.fsPath ?? os.homedir();
+    const template = await launchTemplatePicker(cwd);
+    if (!template) {
+        return;
+    }
+
+    if (template.info.create_dir) {
+        const uri = await window.showSaveDialog({
```
Comment (ManuelHentschel) [inline src/rmarkdown/draft.ts:107]: The `.Rmd` extension might be a bit confusing here. I did not find a really clean solution to configure the SaveDialog to specify a folder, but the following could be used:
``` ts 
        let defaultPath = path.join(cwd, 'draft');
        let i = 1;
        while(fs.existsSync(defaultPath)){
            defaultPath = path.join(cwd, `draft_${++i}`);
        }

        const uri = await window.showSaveDialog({
            defaultUri: Uri.file(defaultPath),
            filters: {
                'Folder': ['']
            },
            saveLabel: 'Create Folder',
            title: 'R Markdown: New Draft'
        });
```

The open dialogue seems to have an option to select only folders, but does not allow to select non-existing folders

File: src/rmarkdown/draft.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    const options: QuickPickOptions = {
+        matchOnDescription: true,
+        matchOnDetail: true,
+        canPickMany: false,
+        ignoreFocusOut: false,
+        placeHolder: '',
+        onDidSelectItem: undefined
+    };
+
+    const items = await getTemplateItems(cwd);
```
Comment (ManuelHentschel) [inline src/rmarkdown/draft.ts:84]: Here we could indicate that a folder needs to be created by e.g.:
``` ts
    for(const item of items){
        if(item.info.create_dir){
            item.label = (item.label || '') + ' $(list-tree) ' ;
        }
    }
```

File: src/rmarkdown/draft.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    const args = [
+        '--silent',
+        '--slave',
+        '--no-save',
+        '--no-restore',
+        '-f',
+        rScriptFile
+    ];
+
+    let items: TemplateItem[] = [];
```
[the remark this answers is not in the document]
Comment (renkun-ken) [reply src/rmarkdown/draft.ts]: Agree

File: src/rmarkdown/draft.ts (hunk 2 above)
Comment (renkun-ken) [reply src/rmarkdown/draft.ts:81]: Didn't know that icons could appear in the item label. `create_dir` is known in `getTemplateItems` when `label` is first specified. `$(new-folder)` might be more direct here.

Thanks for the tip!

File: src/rmarkdown/draft.ts (hunk 1 above)
Comment (renkun-ken) [reply src/rmarkdown/draft.ts:104]: Yes, it is something I'm also considering. I tried `showOpenDialog` with folder-only [options](https://code.visualstudio.com/api/references/vscode-api#OpenDialogOptions), but it didn't work smoothly for saving purpose.

Comment (renkun-ken) [reply src/rmarkdown/draft.ts:104]: Also, `rmarkdown::draft()` will throw error if the target file or folder exists, which I have not handled yet.

Comment (renkun-ken) [reply src/rmarkdown/draft.ts:109]: `rmarkdown::draft()` will create a folder with last file extension removed. For example, `draft.Rmd` will become `draft`; `draft.test.Rmd` will become `draft.test`.

I update the code to check if this folder already exists and let user confirm whether to remove it before creating the draft folder.

**Comment by renkun-ken** [comment]
@ManuelHentschel Any more suggestions?

**Review by ManuelHentschel — APPROVED** [review]
Works nicely, thanks!

**Pull request merged**